File: packages/pofile-ts/src/catalog.ts
/**
 * Catalog conversion helpers for working with simple key-value formats.
 *
 * Provides utilities to convert between a simple catalog format and PO items.
 */

import type { PoItem } from "./types"
import { createItem } from "./Item"
import { parseReference, formatReference, type SourceReference } from "./references"

/** Checks if an object has any own properties (faster than Object.keys().length) */
function hasOwnProperties(obj: object): boolean {
  for (const _ in obj) {
    return true
  }
  return false
}

/**
 * A single entry in the catalog.
 */
export interface CatalogEntry {
  /**
   * The source message (msgid content).
   * Used when the catalog key is a generated ID rather than the source text.
   */
  message?: string

  /**
   * The translated string(s).
   * Use an array for plural forms: [singular, plural, ...]
   * Optional for extraction workflows where translations don't exist yet.
   */
  translation?: string | string[]

  /**
   * Source string for plural forms (msgid_plural).
   * Required when translation is an array.
   */
  pluralSource?: string

  /**
   * Message context for disambiguation (msgctxt).
   */
  context?: string

  /**
   * Translator comments.
   */
  comments?: string[]

  /**
   * Extracted comments (from source code).
   */
  extractedComments?: string[]

  /**
   * Source file references.
   */
  origins?: SourceReference[]

  /**
   * Whether this entry is obsolete.
   */
  obsolete?: boolean

  /**
   * Flags like "fuzzy".
   */
  flags?: Record<string, boolean>
}

/**
 * A catalog is a record of message IDs to their entries.
 */
export type Catalog = Record<string, CatalogEntry>

/**
 * Options for converting catalog to items.
 */
export interface CatalogToItemsOptions {
  /**
   * Include source references in the output.
   * @default true
   */
  includeOrigins?: boolean

  /**
   * Include line numbers in references.
   * @default true
   */
  includeLineNumbers?: boolean

  /**
   * Number of plural forms for the target language.
   * @default 2
   */
  nplurals?: number
}

/**
 * Options for converting items to catalog.
 */
export interface ItemsToCatalogOptions {
  /**
   * Use msgid as the catalog key (true) or use a custom key generator (false).
   * @default true
   */
  useMsgidAsKey?: boolean

  /**
   * Custom function to generate catalog keys from items.
   * Only used when useMsgidAsKey is false.
   */
  keyGenerator?: (item: PoItem) => string

  /**
   * Include origins in the catalog entries.
   * @default true
   */
  includeOrigins?: boolean
}

/** Applies translation to an item */
function applyTranslation(item: PoItem, entry: CatalogEntry): void {
  if (entry.translation === undefined) {
    // No translation yet (extraction workflow)
    item.msgstr = entry.pluralSource ? ["", ""] : [""]
    if (entry.pluralSource) {
      item.msgid_plural = entry.pluralSource
    }
  } else if (Array.isArray(entry.translation)) {
    item.msgstr = entry.translation
    if (entry.pluralSource) {
      item.msgid_plural = entry.pluralSource
    }
  } else {
    item.msgstr = [entry.translation]
  }
}

/** Applies optional fields from entry to item */
function applyOptionalFields(
  item: PoItem,
  entry: CatalogEntry,
  options: { includeOrigins: boolean; includeLineNumbers: boolean }
): void {
  if (entry.context) {
    item.msgctxt = entry.context
  }
  if (entry.comments) {
    item.comments = entry.comments
  }
  if (entry.extractedComments) {
    item.extractedComments = entry.extractedComments
  }
  if (options.includeOrigins && entry.origins) {
    item.references = entry.origins.map((ref) =>
      formatReference(ref, { includeLineNumbers: options.includeLineNumbers })
    )
  }
  if (entry.obsolete) {
    item.obsolete = true
  }
  if (entry.flags) {
    item.flags = { ...entry.flags }
  }
}

/**
 * Converts a catalog to PO items.
 *
 * @example
 * const items = catalogToItems({
 *   "Hello": { translation: "Hallo" },
 *   "greeting": {
 *     message: "Hello {name}",
 *     translation: "Hallo {name}",
 *     context: "informal"
 *   },
 *   "{count} item": {
 *     translation: ["{count} Element", "{count} Elemente"],
 *     pluralSource: "{count} items"
 *   }
 * })
 */
export function catalogToItems(catalog: Catalog, options: CatalogToItemsOptions = {}): PoItem[] {
  const { includeOrigins = true, includeLineNumbers = true, nplurals = 2 } = options

  return Object.entries(catalog).map(([key, entry]) => {
    const item = createItem({ nplurals })
    item.msgid = entry.message ?? key
    applyTranslation(item, entry)
    applyOptionalFields(item, entry, { includeOrigins, includeLineNumbers })
    return item
  })
}

/** Gets the catalog key for an item */
function getCatalogKey(
  item: PoItem,
  useMsgidAsKey: boolean,
  keyGenerator?: (item: PoItem) => string
): string {
  if (useMsgidAsKey) {
    return item.msgid
  }
  if (keyGenerator) {
    return keyGenerator(item)
  }
  return item.msgid
}

/** Adds message field if key differs from msgid */
function addMessageField(
  entry: CatalogEntry,
  item: PoItem,
  key: string,
  useMsgidAsKey: boolean
): void {
  if (!useMsgidAsKey && item.msgid !== key) {
    entry.message = item.msgid
  }
  if (item.msgid_plural) {
    entry.pluralSource = item.msgid_plural
  }
  if (item.msgctxt) {
    entry.context = item.msgctxt
  }
}

/** Adds comments fields to entry if non-empty (handles incomplete items) */
function addCommentsFields(entry: CatalogEntry, item: PoItem): void {
  const comments = item.comments as string[] | undefined
  const extractedComments = item.extractedComments as string[] | undefined
  if (comments && comments.length > 0) {
    entry.comments = comments
  }
  if (extractedComments && extractedComments.length > 0) {
    entry.extractedComments = extractedComments
  }
}

/** Adds metadata fields to an entry (handles incomplete items gracefully) */
function addMetadataFields(entry: CatalogEntry, item: PoItem, includeOrigins: boolean): void {
  addCommentsFields(entry, item)

  const references = item.references as string[] | undefined
  if (includeOrigins && references && references.length > 0) {
    entry.origins = references.map((ref) => parseReference(ref))
  }

  if (item.obsolete) {
    entry.obsolete = true
  }

  const flags = item.flags as Record<string, boolean> | undefined
  if (flags && hasOwnProperties(flags)) {
    entry.flags = { ...flags }
  }
}

/**
 * Converts PO items to a catalog.
 *
 * @example
 * const catalog = itemsToCatalog(items)
 * // → { "Hello": { translation: "Hallo", ... } }
 */
export function itemsToCatalog(items: PoItem[], options: ItemsToCatalogOptions = {}): Catalog {
  const { useMsgidAsKey = true, keyGenerator, includeOrigins = true } = options
  const catalog: Catalog = {}

  for (const item of items) {
    if (!item.msgid) {
      continue
    }

    const key = getCatalogKey(item, useMsgidAsKey, keyGenerator)
    const msgstr = item.msgstr as string[] | undefined
    const entry: CatalogEntry = {
      translation: item.msgid_plural ? (msgstr ?? []) : (msgstr?.[0] ?? "")
    }

    addMessageField(entry, item, key, useMsgidAsKey)
    addMetadataFields(entry, item, includeOrigins)
    catalog[key] = entry
  }

  return catalog
}

/**
 * Merges two catalogs, with the second catalog taking precedence.
 *
 * Useful for merging extracted messages with existing translations.
 *
 * @example
 * const merged = mergeCatalogs(existingCatalog, newCatalog)
 */
export function mergeCatalogs(base: Catalog, updates: Catalog): Catalog {
  const merged: Catalog = {}

  // Copy base entries
  for (const [key, entry] of Object.entries(base)) {
    merged[key] = { ...entry }
  }

  // Merge updates
  for (const [key, update] of Object.entries(updates)) {
    const existing = merged[key]
    if (existing) {
      // Merge with existing entry
      merged[key] = {
        ...existing,
        ...update,
        // Merge arrays instead of replacing
        comments: update.comments ?? existing.comments,
        extractedComments: update.extractedComments ?? existing.extractedComments,
        origins: update.origins ?? existing.origins,
        flags: { ...existing.flags, ...update.flags }
      }
    } else {
      merged[key] = { ...update }
    }
  }

  return merged
}

File: crates/pofile/src/catalog.rs
//! Catalog conversion helpers.

use std::collections::BTreeMap;

use crate::po::PoItem;
use crate::references::{
    format_reference, parse_reference, FormatReferenceOptions, ReferenceError, SourceReference,
};

/// A single catalog entry.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct CatalogEntry {
    /// Source message when the catalog key is not the `msgid`.
    pub message: Option<String>,
    /// Translation value or plural variants.
    pub translation: Option<CatalogTranslation>,
    /// Plural source string.
    pub plural_source: Option<String>,
    /// Message context.
    pub context: Option<String>,
    /// Translator comments.
    pub comments: Option<Vec<String>>,
    /// Extracted comments.
    pub extracted_comments: Option<Vec<String>>,
    /// Source references.
    pub origins: Option<Vec<SourceReference>>,
    /// Whether the entry is obsolete.
    pub obsolete: Option<bool>,
    /// Flags such as `fuzzy`.
    pub flags: Option<BTreeMap<String, bool>>,
}

/// Translation payload stored inside a [`CatalogEntry`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CatalogTranslation {
    /// Singular translation.
    Singular(String),
    /// Plural translations.
    Plural(Vec<String>),
}

/// Catalog keyed by message ID or another caller-defined key.
pub type Catalog = BTreeMap<String, CatalogEntry>;

/// Options for converting a catalog to PO items.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CatalogToItemsOptions {
    /// Include origin references in the converted items.
    pub include_origins: bool,
    /// Include line numbers when formatting references.
    pub include_line_numbers: bool,
    /// Number of plural forms to initialize for new items.
    pub nplurals: usize,
}

impl Default for CatalogToItemsOptions {
    fn default() -> Self {
        Self {
            include_origins: true,
            include_line_numbers: true,
            nplurals: 2,
        }
    }
}

/// Options for converting items back into a catalog.
pub struct ItemsToCatalogOptions<'a> {
    /// Keep `msgid` as the catalog key.
    pub use_msgid_as_key: bool,
    /// Include parsed references as `origins`.
    pub include_origins: bool,
    /// Optional key generator used when `use_msgid_as_key` is `false`.
    pub key_generator: Option<&'a dyn Fn(&PoItem) -> String>,
}

impl Default for ItemsToCatalogOptions<'_> {
    fn default() -> Self {
        Self {
            use_msgid_as_key: true,
            include_origins: true,
            key_generator: None,
        }
    }
}

/// Convert a catalog into PO items.
#[must_use]
pub fn catalog_to_items(catalog: &Catalog, options: CatalogToItemsOptions) -> Vec<PoItem> {
    catalog
        .iter()
        .map(|(key, entry)| {
            let mut item = PoItem::new(options.nplurals);
            item.msgid = entry.message.clone().unwrap_or_else(|| key.clone());

            apply_translation(&mut item, entry);
            apply_optional_fields(&mut item, entry, options);
            item
        })
        .collect()
}

/// Convert PO items into a catalog.
pub fn items_to_catalog(
    items: &[PoItem],
    options: ItemsToCatalogOptions<'_>,
) -> Result<Catalog, ReferenceError> {
    let mut catalog = Catalog::new();

    for item in items {
        if item.msgid.is_empty() {
            continue;
        }

        let key = get_catalog_key(item, &options);
        let mut entry = CatalogEntry {
            translation: Some(if item.msgid_plural.is_some() {
                CatalogTranslation::Plural(item.msgstr.clone())
            } else {
                CatalogTranslation::Singular(item.msgstr.first().cloned().unwrap_or_default())
            }),
            ..CatalogEntry::default()
        };

        add_message_field(&mut entry, item, &key, options.use_msgid_as_key);
        add_metadata_fields(&mut entry, item, options.include_origins)?;
        catalog.insert(key, entry);
    }

    Ok(catalog)
}

/// Merge two catalogs, preferring values from `updates`.
#[must_use]
pub fn merge_catalogs(base: &Catalog, updates: &Catalog) -> Catalog {
    let mut merged = base.clone();

    for (key, update) in updates {
        if let Some(existing) = merged.get_mut(key) {
            existing.message = update.message.clone().or_else(|| existing.message.clone());
            existing.translation = update
                .translation
                .clone()
                .or_else(|| existing.translation.clone());
            existing.plural_source = update
                .plural_source
                .clone()
                .or_else(|| existing.plural_source.clone());
            existing.context = update.context.clone().or_else(|| existing.context.clone());
            existing.comments = update
                .comments
                .clone()
                .or_else(|| existing.comments.clone());
            existing.extracted_comments = update
                .extracted_comments
                .clone()
                .or_else(|| existing.extracted_comments.clone());
            existing.origins = update.origins.clone().or_else(|| existing.origins.clone());
            existing.obsolete = update.obsolete.or(existing.obsolete);

            let mut flags = existing.flags.clone().unwrap_or_default();
            if let Some(update_flags) = &update.flags {
                flags.extend(update_flags.clone());
            }
            existing.flags = (!flags.is_empty()).then_some(flags);
        } else {
            merged.insert(key.clone(), update.clone());
        }
    }

    merged
}

fn apply_translation(item: &mut PoItem, entry: &CatalogEntry) {
    match &entry.translation {
        None => {
            item.msgstr = if entry.plural_source.is_some() {
                vec![String::new(), String::new()]
            } else {
                vec![String::new()]
            };
            item.msgid_plural = entry.plural_source.clone();
        }
        Some(CatalogTranslation::Singular(text)) => item.msgstr = vec![text.clone()],
        Some(CatalogTranslation::Plural(texts)) => {
            item.msgstr = texts.clone();
            item.msgid_plural = entry.plural_source.clone();
        }
    }
}

fn apply_optional_fields(item: &mut PoItem, entry: &CatalogEntry, options: CatalogToItemsOptions) {
    item.msgctxt = entry.context.clone();

    if let Some(comments) = &entry.comments {
        item.comments = comments.clone();
    }
    if let Some(comments) = &entry.extracted_comments {
        item.extracted_comments = comments.clone();
    }
    if options.include_origins {
        if let Some(origins) = &entry.origins {
            item.references = origins
                .iter()
                .map(|reference| {
                    format_reference(
                        reference,
                        FormatReferenceOptions {
                            include_line_numbers: options.include_line_numbers,
                        },
                    )
                })
                .collect();
        }
    }
    if entry.obsolete.unwrap_or(false) {
        item.obsolete = true;
    }
    if let Some(flags) = &entry.flags {
        item.flags = flags.clone();
    }
}

fn get_catalog_key(item: &PoItem, options: &ItemsToCatalogOptions<'_>) -> String {
    if options.use_msgid_as_key {
        item.msgid.clone()
    } else if let Some(key_generator) = options.key_generator {
        key_generator(item)
    } else {
        item.msgid.clone()
    }
}

fn add_message_field(entry: &mut CatalogEntry, item: &PoItem, key: &str, use_msgid_as_key: bool) {
    if !use_msgid_as_key && item.msgid != key {
        entry.message = Some(item.msgid.clone());
    }
    entry.plural_source = item.msgid_plural.clone();
    entry.context = item.msgctxt.clone();
}

fn add_metadata_fields(
    entry: &mut CatalogEntry,
    item: &PoItem,
    include_origins: bool,
) -> Result<(), ReferenceError> {
    if !item.comments.is_empty() {
        entry.comments = Some(item.comments.clone());
    }
    if !item.extracted_comments.is_empty() {
        entry.extracted_comments = Some(item.extracted_comments.clone());
    }

    if include_origins && !item.references.is_empty() {
        entry.origins = Some(
            item.references
                .iter()
                .map(|reference| parse_reference(reference))
                .collect::<Result<Vec<_>, _>>()?,
        );
    }

    if item.obsolete {
        entry.obsolete = Some(true);
    }
    if !item.flags.is_empty() {
        entry.flags = Some(item.flags.clone());
    }

    Ok(())
}
